feat(chart): create bare entrypoint to reduce bundle size

Theeraphat-Sorasetsakul opened the pull request:
## Description

Provide new entrypoint of ef-chart for app to optimize size of bundle

Fixes # ([STG-421](https://jira.refinitiv.com/browse/STG-421))

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

Commit:
feat(chart): create bare entrypoint to reduce bundle size

wsuwt commented on packages/elements/src/chart/bare.ts:
```diff
@@ -0,0 +1,620 @@
+   * Html canvas element
+   * @type {HTMLCanvasElement}
+   */
+  @query('canvas')
```
Can we use `ref` instead ?

Theeraphat-Sorasetsakul replied:
Done.

Commit:
refactor(chart): update canvas pointer to using ref

Commit:
chore(chart): unregister chartjs features to be bare

wsuwt committed:
refactor(chart): chart bare entry PR (#691)

* test: increase mocha test timeout to 5 sec (#642)

* refactor(combo-box): use conditional template rendering instead of `hidden` attribute to hide clear icon (#602)

* fix(calendar): unexpected `blur` event fired when view changes (#655)

* chore(release): publish [skip ci]

 - @refinitiv-ui/docs@6.4.10
 - @refinitiv-ui/elements@6.7.3

* docs: fix wrong item type of multiple list demo (#666)

* fix: add checkbox theme to item element to made it render on multiple mode

* fix: revert import checkbox theme

* docs: fix wrong item type of multiple list demo

* docs: fix generated `tree-item` type in demo page

---------

Co-authored-by: Sarin Udompanish <sarin.udompanish2@lseg.com>
Co-authored-by: wsuwt <wasuwat.limsuparhat2@lseg.com>

* chore(deps): bump yaml from 2.2.1 to 2.2.2 (#664)

Bumps [yaml](https://github.com/eemeli/yaml) from 2.2.1 to 2.2.2.
- [Release notes](https://github.com/eemeli/yaml/releases)
- [Commits](https://github.com/eemeli/yaml/compare/v2.2.1...v2.2.2)

---
updated-dependencies:
- dependency-name: yaml
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Wasuwat Limsuparhat <86233706+wsuwt@users.noreply.github.com>

* docs: notify user to reload page after switch theme (#670)

* docs: notify user to reload page after switch theme

* Update documents/src/pages/build-app/theme-switching.md

Co-authored-by: Wasuwat Limsuparhat <86233706+wsuwt@users.noreply.github.com>

---------

Co-authored-by: Wasuwat Limsuparhat <86233706+wsuwt@users.noreply.github.com>

* chore: add black duck workflow (#667)

* feat: remove font bold (#665)

* feat: remove font bold

* fix: native header element font-weight

* docs: fix font-weight in demo pages

* chore: update LICENSE removing `bold` font reference

* fix: remove font bold from elements

---------

Co-authored-by: Sarin Udompanish <sarin.udompanish2@lseg.com>
Co-authored-by: wsuwt <wasuwat.limsuparhat2@lseg.com>
Co-authored-by: Wasuwat Limsuparhat <86233706+wsuwt@users.noreply.github.com>

* feat: export `nothing` value from `lit` (#671)

* chore(release): publish [skip ci]

 - @refinitiv-ui/docs@6.5.0
 - @refinitiv-ui/core@6.4.0
 - @refinitiv-ui/demo-block@6.1.0
 - @refinitiv-ui/elements@6.8.0
 - @refinitiv-ui/halo-theme@6.5.0
 - @refinitiv-ui/translate@6.0.18

* docs: add file extension to dynamic import of theme switching example (#683)

* test: convert resize observer loop error to warning (#679)

* feat(test-helper): add mocha util converting resize observer loop error to warning

* test(canvas): apply observer loop handler

* test(clock): apply observer loop handler

* test(heatmap): apply observer loop handler

* test(interactive-chart): apply observer loop handler

* test(layout): apply observer loop handler

* test(notification): apply observer loop handler

* test(overlay): apply observer loop handler

* test(sparkline): apply observer loop handler

* test(swing-gauge): apply observer loop handler

* test(tornado-chart): apply observer loop handler

* test(test-helpers): add test for resize observer loop handler

* refactor(test-helpers): refactor resize observer loop handler to test helper's side effect

* test(elements): remove resize observer loop handler import

* refactor(test-helpers): cleanup export

* refactor(test-helpers): fix code smell

---------

Co-authored-by: Wasuwat Limsuparhat <86233706+wsuwt@users.noreply.github.com>

* fix: responsive element not work in ios 15 and below (#680)

Co-authored-by: wattachai <117723407+wattachai-lseg@users.noreply.github.com>

* chore(release): publish [skip ci]

 - @refinitiv-ui/docs@6.5.1
 - @refinitiv-ui/core@6.4.1
 - @refinitiv-ui/demo-block@6.1.1
 - @refinitiv-ui/elements@6.8.1
 - @refinitiv-ui/i18n@6.0.12
 - @refinitiv-ui/test-helpers@6.0.9
 - @refinitiv-ui/translate@6.0.19

* test: fix to using web/test-runner syntax

* test: fix ignore coverage syntax

* chore: replace web dev server with vite (#681)

* chore: update demo page

* chore: remove all web-dev-server config and update demo

* chore: add missing demo css file to heatmap demo

* chore: remove native button demo

* chore: change ignore eslint in demo block

* chore: remove visibility hidden from demo block

---------

Co-authored-by: Sarin-Udompanish <sarin.udompanish2@lseg.com>

* refactor: chart bare entry PR

* test: update chart snapshot

* docs(chart): revert Type back to support jsdoc

---------

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: wattachai <117723407+wattachai-lseg@users.noreply.github.com>
Co-authored-by: refinitiv-ui-dev <RefinitivUIDev@refinitiv.com>
Co-authored-by: Sarin-Udompanish <86759822+Sarin-Udompanish@users.noreply.github.com>
Co-authored-by: Sarin Udompanish <sarin.udompanish2@lseg.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Napat Bualoy <91940544+bualoy-napat@users.noreply.github.com>
Co-authored-by: Theeraphat Sorasetsakul <Theeraphat.Sorasetsakul@refinitiv.com>
Co-authored-by: Theeraphat-Sorasetsakul <86758473+Theeraphat-Sorasetsakul@users.noreply.github.com>

Commit:
Merge remote-tracking branch 'origin/v7' into chart/new-entrypoint

Commit:
refactor(chart): restructure file to make it cleaner

Commit:
docs(chart): rewrite to make it understandable

Commit:
Merge branch 'migrate-chartjs4' into chart/new-entrypoint

Commit:
docs(chart): remove unwanted line break

wsuwt commented on packages/elements/src/chart/elements/chart.ts:
```diff
@@ -0,0 +1,603 @@
+   * Set global configuration of ChartJS
+   * @returns {void}
+   */
+  // TODO: Try and remove the need for global object modification.
```
Is this TODO still valid ? If it is, is it worth trying out ?

Theeraphat-Sorasetsakul replied:
I tried and found that without global config cannot cover our style to creating a new custom feature, for example a new custom scale. And also with global config is easier readability.

wsuwt replied:
So that means this TODO is not valid anymore. Should we remove it ?

Theeraphat-Sorasetsakul replied:
You're right. I will remove it.

wsuwt commented on packages/elements/src/chart/elements/chart.ts:
```diff
@@ -0,0 +1,603 @@
+      let backgroundColor;
+      const isMultipleDatasets = (chart.config.data.datasets.length > 1);
+      // From old requirement, Only line, radar, scatter, polarArea type are opaque backgroundColor
+      switch (dataset.type || this.config?.type) {
```
```suggestion
      switch (dataset.type ?? this.config?.type) {
```

wsuwt replied:
This only check for `null` and `undefined` though.

Commit:
docs(chart): add missing comma

wsuwt commented on packages/elements/src/chart/elements/chart.ts:
```diff
@@ -0,0 +1,603 @@
+    }
+    legends.forEach((legend, i) => {
+      legend.lineWidth = Number(datasets[i].borderWidth) || 0;
+      switch (datasets[i].type || chartType) {
```
```suggestion
      switch (datasets[i].type ?? chartType) {
```

Theeraphat-Sorasetsakul committed:
Update packages/elements/src/chart/elements/chart.ts

Co-authored-by: Wasuwat Limsuparhat <86233706+wsuwt@users.noreply.github.com>

Theeraphat-Sorasetsakul committed:
Update packages/elements/src/chart/elements/chart.ts

Co-authored-by: Wasuwat Limsuparhat <86233706+wsuwt@users.noreply.github.com>

Commit:
docs(chart): remove TODO comment

Commit:
Merge branch 'chart/new-entrypoint' of github.com:Refinitiv/refinitiv-ui into chart/new-entrypoint

wsuwt commented on packages/elements/src/chart/elements/chart.ts:
```diff
@@ -0,0 +1,603 @@
+}
+
+/**
+ * Charting component that use ChartJS library
```
```suggestion
 * Charting component that uses ChartJS library
```

Theeraphat-Sorasetsakul committed:
Update packages/elements/src/chart/elements/chart.ts

Co-authored-by: Wasuwat Limsuparhat <86233706+wsuwt@users.noreply.github.com>

wattachai-lseg commented on packages/elements/src/chart/plugins/index.ts:
```diff
@@ -0,0 +1,4 @@
+import doughnutCenterLabelPlugin from './doughnut-center-label.js';
+export {
+  doughnutCenterLabelPlugin
+};
```
```suggestion
export * from './doughnut-center-label.js';
```

wattachai-lseg commented on packages/elements/src/chart/plugins/doughnut-center-label.ts:
change `export default plugins;` to `export { plugins as doughnutCenterLabelPlugin };`

Theeraphat-Sorasetsakul committed:
Update packages/elements/src/chart/plugins/index.ts

Co-authored-by: wattachai <117723407+wattachai-lseg@users.noreply.github.com>

wattachai-lseg committed:
docs(chart): revise reduce bundle size section

Commit:
refactor(chart): fix export way

wsuwt commented on packages/elements/src/chart/elements/chart.ts:
```diff
@@ -0,0 +1,600 @@
+   * @returns The value as a number, or, undefined if NaN.
+   */
+  protected cssVarAsNumber (...args: string[]): number | undefined {
+    const result = Number(this.getComputedVariable(...args).replace(/\D+$/, ''));
```
Can you double check this security hotspot ? https://sonarcloud.io/project/security_hotspots?id=refinitiv_refinitiv-ui&pullRequest=687&resolved=false&types=SECURITY_HOTSPOT

Theeraphat-Sorasetsakul replied:
We can be continuous on [STG-434](https://jira.refinitiv.com/browse/STG-434).

Commit:
docs(chart): add more examples

Commit:
fix(chart): prevent vulnerable code

Commit:
docs(chart): fix wording to be readable

Commit:
fix(chart): fix regex to be vulnerable
